chain/types/actor_event.go

```diff
@@ -0,0 +1,70 @@
+package types
+
+import (
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+)
+
+type ActorEventBlock struct {
+	// The value codec to match when filtering event values.
+	Codec uint64 `json:"codec"`
+
+	// The value to want to match on associated with the corresponding "event key"
+	// when filtering events.
+	// Should be a byte array encoded with the specified codec.
+	// Assumes base64 encoding when converting to/from JSON strings.
+	Value []byte `json:"value"`
+}
+
+type SubActorEventFilter struct {
+	Filter ActorEventFilter `json:"filter"`
+
+	// If true, all available matching historical events will be written to the response stream
+	// before any new real-time events that match the given filter are written.
+	Prefill bool `json:"prefill"`
+}
+
+type ActorEventFilter struct {
+	// Matches events from one of these actors, or any actor if empty.
+	// For now, this MUST be a Filecoin address.
+	Addresses []address.Address `json:"address"`
+
+	// Matches events with the specified key/values, or all events if empty.
+	// If the value is an empty slice, the filter will match on the key only, accepting any value.
+	Fields map[string][]ActorEventBlock `json:"fields"`
+
+	// Interpreted as an epoch (in hex) or one of "latest" for last mined block, "earliest" for first,
+	// Optional, default: "latest".
```

Comment:
by default, the call is only "give me events from the current epoch" isn't it? I wonder if different defaults makes sense for `Get` vs `Subscribe`?

Reply:
What defaults do you have in mind ? My suggestion would be defer to actual user feedback here and make changes in a subsequent release once people have tried this out. F

or now, I've just picked this up from the existing ETH events API which has been running the wild for some time.

Reply:
I'm mainly wondering, from a user perspective, what the expectation with a call like `GetActorEvents` is? The current implementation just gives them any for this current epoch. Maybe that's reasonable and more than that would be way too much. I'm tempted to think that users will come to this imagining that the default filter is "just give me everything!". But in fact it's actually "just give me everything .. in the current epoch". For `Subscribe` I always think of those kinds of APIs in terms of `tail -f`, where fetching historical stuff _is_ something you go out of your way to do. Then we have the doubling up of `Prefill` for `Subscribe`, so even if we set this default to `"earliest"` for both and `Prefill` remains `false` then you'd get the `tail -f` behaviour for default `Subscribe` and the maximal output with default `Get`.

My analogies are something like:

```
# "Get" system logs - all the things!
cat /var/log/syslog
# Follow / "Subscribe" system logs - only the new stuff
tail -f /var/log/syslog
# "Get" system logs - with filters applied
cat /var/log/syslog | grep '^boop:\|ERROR'
```

Not a big deal, we can always change it later. Although it'd be quite a surprise change in default to go from minimal to maximal.